components/script/dom/transformstream.rs

```diff
@@ -0,0 +1,950 @@
+        // Let cancelAlgorithm be an algorithm which returns a promise resolved with undefined.
+
+        // If transformerDict["transform"] exists, set transformAlgorithm to an algorithm which
+        // takes an argument
+        // chunk and returns the result of invoking transformerDict["transform"] with argument
+        // list « chunk, controller »
+        // and callback this value transformer.
+
+        // If transformerDict["flush"] exists, set flushAlgorithm to an algorithm which returns
+        // the result
+        // of invoking transformerDict["flush"] with argument list « controller » and callback
+        // this value transformer.
+
+        // If transformerDict["cancel"] exists, set cancelAlgorithm to an algorithm which takes an argument
+        // reason and returns the result of invoking transformerDict["cancel"] with argument list « reason »
+        // and callback this value transformer.
+        controller.set_transform_obj(transformer_obj);
+
+        // Perform ! SetUpTransformStreamDefaultController(stream, controller,
+        // transformAlgorithm, flushAlgorithm, cancelAlgorithm).
+        self.set_up_transform_stream_default_controller(&controller);
+    }
+
+    /// <https://streams.spec.whatwg.org/#transform-stream-default-sink-write-algorithm>
+    pub(crate) fn transform_stream_default_sink_write_algorithm(
+        &self,
+        cx: SafeJSContext,
+        global: &GlobalScope,
+        chunk: SafeHandleValue,
+        can_gc: CanGc,
+    ) -> Fallible<Rc<Promise>> {
+        // Assert: stream.[[writable]].[[state]] is "writable".
+        assert!(self.writable.get().is_some());
+
+        // Let controller be stream.[[controller]].
+        let controller = self.controller.get().expect("controller is not set");
+
+        // If stream.[[backpressure]] is true,
+        if self.backpressure.get() {
+            // Let backpressureChangePromise be stream.[[backpressureChangePromise]].
+            let backpressure_change_promise = self.backpressure_change_promise.borrow();
+
+            // Assert: backpressureChangePromise is not undefined.
+            assert!(backpressure_change_promise.is_some());
+
+            // Return the result of reacting to backpressureChangePromise with the following fulfillment steps:
+            let result_promise = Promise::new(global, can_gc);
+            rooted!(in(*cx) let mut fulfillment_handler = Some(TransformBackPressureChangePromiseFulfillment {
+                controller: Dom::from_ref(&controller),
+                writable: Dom::from_ref(&self.writable.get().expect("writable stream")),
+                chunk: Heap::boxed(chunk.get()),
+                result_promise: result_promise.clone(),
+            }));
+
+            let handler = PromiseNativeHandler::new(
+                global,
+                fulfillment_handler.take().map(|h| Box::new(h) as Box<_>),
+                None,
```

Comment:
Here you also need a rejection handler, which should just pass the rejection through to `result_promise`(otherwise you never reject it).